dirty_cat/_deduplicate.py

```diff
@@ -0,0 +1,175 @@
+"""
+Implements deduplication based on clustering string distance matrices.
+This works best if there is a number of underlying categories that
+sometimes appear in the data with small variations and/or misspellings.
+"""
+from typing import List, Literal, Optional, Sequence, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from scipy.cluster.hierarchy import fcluster, linkage
+from scipy.spatial.distance import pdist, squareform
+from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
+from sklearn.metrics import silhouette_score
+
+
+def compute_ngram_distance(
+    unique_words: Union[Sequence[str], np.ndarray],
+    ngram_range: Tuple[int, int] = (2, 4),
+    analyzer: str = "char_wb",
+) -> np.ndarray:
+    """Computes the condensed n-gram distance matrix between words in
+    `unique_words`, using the
+    term frequency-inverse document frequency (tf-idf).
+
+    Parameters
+    ----------
+    unique_words : Union[Sequence[str], np.ndarray]
+        Sequence or array of unique words from the original data.
+    ngram_range : Tuple[int, int], optional
+        The n-gram range to compute the distance in, by default (2, 4)
+    analyzer : str, optional
+        Analyzer to extract n-grams, by default "char_wb"
+
+    Returns
+    -------
+    np.ndarray
+        An n-times-(n-1)/2 matrix of n-gram TfIdf distances between `unique_words`.
+    """
+    enc = CountVectorizer(ngram_range=ngram_range, analyzer=analyzer)
+    encoded = TfidfTransformer().fit_transform(enc.fit_transform(unique_words))
```

Comment:
This is equivalent to using scikit-learn's [TDIDF Vectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html).